ledger/src/slot_stats.rs

```diff
@@ -0,0 +1,90 @@
+use {
+    crate::blockstore_meta::SlotMeta, bitflags::bitflags, lru::LruCache, solana_sdk::clock::Slot,
+};
+
+const SLOTS_STATS_CACHE_CAPACITY: usize = 300;
+
+macro_rules! get_mut_entry (
+    ($cache:expr, $key:expr) => (
+        match $cache.get_mut(&$key) {
+            Some(entry) => entry,
+            None => {
+                $cache.put($key, SlotStats::default());
+                $cache.get_mut(&$key).unwrap()
+            }
+        }
+    );
+);
+
+#[derive(Copy, Clone, Debug)]
+pub(crate) enum ShredSource {
+    Turbine,
+    Repaired,
+    Recovered,
+}
+
+bitflags! {
+    #[derive(Default)]
+    struct SlotFlags: u8 {
+        const DEAD   = 0b00000001;
+        const FULL   = 0b00000010;
+        const ROOTED = 0b00000100;
+    }
+}
+
+#[derive(Default)]
+struct SlotStats {
+    flags: SlotFlags,
+    num_repaired: usize,
+    num_recovered: usize,
+}
+
+pub(crate) struct SlotsStats(LruCache<Slot, SlotStats>);
+
+impl Default for SlotsStats {
+    fn default() -> Self {
+        // LruCache::unbounded because capacity is enforced manually.
+        Self(LruCache::unbounded())
+    }
+}
+
+impl SlotsStats {
+    pub(crate) fn add_shred(&mut self, slot: Slot, source: ShredSource) {
+        let entry = get_mut_entry!(self.0, slot);
+        match source {
+            ShredSource::Turbine => (),
+            ShredSource::Repaired => entry.num_repaired += 1,
+            ShredSource::Recovered => entry.num_recovered += 1,
+        }
+        self.maybe_evict_cache();
+    }
+
+    pub(crate) fn set_full(&mut self, slot_meta: &SlotMeta) {
+        let total_time_ms =
+            solana_sdk::timing::timestamp().saturating_sub(slot_meta.first_shred_timestamp);
+        let last_index = slot_meta
+            .last_index
+            .and_then(|ix| i64::try_from(ix).ok())
+            .unwrap_or(-1);
+        let entry = get_mut_entry!(self.0, slot_meta.slot);
+        if !entry.flags.contains(SlotFlags::FULL) {
+            datapoint_info!(
+                "shred_insert_is_full",
+                ("total_time_ms", total_time_ms, i64),
+                ("slot", slot_meta.slot, i64),
+                ("last_index", last_index, i64),
+                ("num_repaired", entry.num_repaired, i64),
+                ("num_recovered", entry.num_recovered, i64),
+            );
+        }
+        entry.flags |= SlotFlags::FULL;
+        self.maybe_evict_cache();
+    }
+
+    fn maybe_evict_cache(&mut self) {
+        while self.0.len() > SLOTS_STATS_CACHE_CAPACITY {
```

Comment:
This seems fine because slot slot full should always come before slot rooted, in this pipeline of events:

 optimistic execution signal -> slot full -> slot rooted

we'll never get a scenario where we detect optimistic execution, but miss that the slot was rooted and falsely report an error.

Reply:
> optimistic execution signal -> slot full

I believe "slot full" is only with respect to _data_ shreds.
and, number of shreds to signal OE > number of shreds to recover the batch (= slot full)

So, "slot is full" should happen before OE signal. no?

But either way, I am not sure why that matters here?

This is only for metrics (for all slots), and probably we should not reuse the same metrics constructs to infer OE signal; here I am just trying to:
* avoid the worst case scenario where the validator does not make root, and this hash-map grows boundlessly.
* give enough time to all slots (whether rooted, dead, or otherwise) to ingest shreds from turbine before submitting metrics.

Reply:
> I believe "slot full" is only with respect to data shreds.

Ah yup, that's right 

> But either way, I am not sure why that matters here?

I was thinking this could be a convenient way for @jbiseda to implement his experiment to figure out the number of OE slots that are *not* rooted since we were tracking the state here for roots.

Reply:
> I was thinking this could be a convenient way for @jbiseda to implement his experiment to figure out the number of OE slots that are _not_ rooted since we were tracking the state here for roots.

This change was indeed intended to ease the path to obtain some _initial lightweight metrics_ for number of shreds received in each erasure batch and if the slot was ultimately rooted or not.

However, I believe the actual final implementation of the OE signal (not the metrics), requires to be a lot more conservative and do more sanity checks, which might require persisting some data into blockstore e.g. in slot-meta or elsewhere.

Reply:
got it, thanks!